File: store-app/src/main/java/com/comp5348/store/service/ProductService.java
package com.comp5348.store.service;

import com.comp5348.store.dto.ProductDTO;
import com.comp5348.store.entity.Product;

import java.util.List;

public interface ProductService {

    // Create product
    ProductDTO createProduct(ProductDTO productDTO);

    // Update product
    ProductDTO updateProduct(Long id, ProductDTO productDTO);

    // Delete product
    void deleteProduct(Long id);

    // Search product by id
    ProductDTO getProductById(Long id);

    // Search all active products
    List<ProductDTO> getAllActiveProducts();

    // Search products by category
    List<ProductDTO> getProductsByCategory(String category);

    // Search products
    List<ProductDTO> searchProducts(String keyword);

    // Get all categories
    List<String> getAllCategories();

    // Check whether the goods exist
    boolean productExists(Long id);
}

File: store-frontend/src/services/productService.ts
import api from './api';
import { Product } from '../types';

export const productService = {
  getAllProducts: async (): Promise<Product[]> => {
    const response = await api.get('/products');
    return response.data;
  },

  getProductById: async (id: number): Promise<Product> => {
    const response = await api.get(`/products/${id}`);
    return response.data;
  },

  getProductsByCategory: async (category: string): Promise<Product[]> => {
    const response = await api.get(`/products/category/${category}`);
    return response.data;
  },

  searchProducts: async (keyword: string): Promise<Product[]> => {
    const response = await api.get(`/products/search?keyword=${encodeURIComponent(keyword)}`);
    return response.data;
  },

  getAllCategories: async (): Promise<string[]> => {
    const response = await api.get('/products/categories');
    return response.data;
  },

  checkProductExists: async (id: number): Promise<boolean> => {
    const response = await api.get(`/products/${id}/exists`);
    return response.data;
  },
  
  getProductTotalStock: async (id: number): Promise<number> => {
    const response = await api.get(`/products/${id}/stock`);
    return response.data;
  },
};

